### PR TITLE
Issue 5313: Only upload font vertex buffer data at the end of a draw command

### LINKED ISSUE #5313: Performance regression when rendering text
**Describe the bug (REQUIRED)**
I noticed that the fix for [this issue](https://github.com/defold/defold/issues/4901) was reverted in [this commit](https://github.com/defold/defold/commit/6fc9dc26fce06c3c85986ed671c134061cd61e73). This severely impacts text rendering performance, especially in Domains of Dusk, which is very text-heavy. I'm curious: Was the fix breaking something else?

This is on an Intel i7 8700, with the built-in iGPU (UHD 630). No other text apart from the one visible on screen is being rendered:
Screenshot 2020-11-05 at 12 02 01

**To Reproduce (REQUIRED)**
1. Render text nodes

**Expected behavior (REQUIRED)**
Text shouldn't be so resource-intensive.

**Defold version (REQUIRED):**
  - 1.2.174

**Platforms (REQUIRED):**
 - Platforms: macOS
 - OS: 10.15.7

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

**Comment by JCash** [comment]
Screenshot 2021-02-26 at 12 52 52

>FontCharacterCount  3556    
FontVertexBuffer    2389632

File: engine/render/src/render/font_renderer.cpp [hunk 1]
```diff
@@ -1038,26 +1038,36 @@ namespace dmRender
         HRenderContext render_context = (HRenderContext)params.m_UserData;
         TextContext& text_context = render_context->m_TextContext;
 
+        // This function is called for both game object text (labels) and gui
+        // Once the sprites are done rendering, it is ok to reuse/reset the state
+        // and the gui can start rendering its text
+        // See also ClearRenderObjects() in render.cpp for clearing the text entries
+        if (text_context.m_Frame != text_context.m_PreviousFrame)
```
Comment (JCash) [inline engine/render/src/render/font_renderer.cpp:1045]: It's better to let the render engine decide when a new frame occurs.
We already had this concept for the same reason.

File: engine/render/src/render/font_renderer.cpp [hunk 2]
```diff
@@ -1038,26 +1038,36 @@ namespace dmRender
...
-                    }
-                    DM_COUNTER("FontVertexBuffer", buffer_size);
+                    dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, 0, 0, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+                    dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, buffer_size, text_context.m_ClientBuffer, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+
+                    uint32_t num_vertices = text_context.m_VertexIndex - text_context.m_VerticesFlushed;
+                    text_context.m_VerticesFlushed = text_context.m_VertexIndex;
+
+                    DM_COUNTER("FontCharacterCount", num_vertices / 6); // each quad is two triangles
+                    DM_COUNTER("FontVertexBuffer", num_vertices * sizeof(GlyphVertex));
```
Comment (JCash) [inline engine/render/src/render/font_renderer.cpp:1070]: The numbers add up.
Added actual character count as well.

**Review by britzl — APPROVED** [review]
Looks good!

**Pull request merged**